api.bs

```diff
@@ -142,8 +167,147 @@ The aggregation service:
 
 # API Details # {#api}
 
+Open questions:
+*   Filter/query language
+*   Reports are sent to aggregation system directly, or via conversion site? Or
+    option of either? => via conversion site
+*   Epochs
+
 TODO
 
+## ListAggregationSystems API ## {#list-aggregation-systems-api}
+
+navigator.privateAttribution.listAggregationSystems()
+
+<xmp class=idl>
+dictionary PrivateAttributionAggregationSystem {
+  required DOMString id;
+};
+</xmp>
+
+## SaveImpression API ## {#save-impression-api}
+
+<pre>
+navigator.privateAttribution.saveImpression({
+  type: "view",                  // either "view" or "click"
+  index: 3,                      // the histogram index for counting this impression
+  ad: "sample-campaign-eijb",    // a unique identifier for the ad placement
+  target: "advertiser.example",  // the advertiser site where a conversion will occur
+});
+</pre>
+
+Add:
+* attribution system
+* TTL
+* DP parameters
+
+Questions:
+*   Revisit the set of impression types. Can we get rid of it, and put it in the
+    ad ID? Or generalize to "attribution constraint"?
+
+<xmp class=idl>
+enum PrivateAttributionImpressionType { "view", "click" };
+
+dictionary PrivateAttributionImpressionOptions {
+  PrivateAttributionImpressionType type = "view";
+  required unsigned long index;
+  required DOMString ad;
+  required DOMString target;
+};
+
+[SecureContext, Exposed=Window]
+interface PrivateAttribution {
+  [Throws] undefined saveImpression(DOMString aggregationSystemId, PrivateAttributionImpressionOptions options);
+};
+</xmp>
+
+Implicit saveImpression API inputs:
+* Timestamp (epoch?)
+* Source site
+
+
+### Operation ### {#save-impression-api-operation}
+
+1. Validate inputs
+2. If the private attribution API is not enabled, discard the impression data.
+3. Save the impression to the store.
+
+
+## MeasureConversion API ## {#measure-conversion-api}
+
+TODO:
+* Add conversion value
+* Change filter data
+
+
+navigator.privateAttribution.measureConversion({
+  // the number of buckets in the histogram
+  "size": 20,
+
+  // only consider impressions within the last N days
+  lookbackDays: 30,
+  // the type of impression to match against (if omitted, match either)
+  impression: "view",
+  // a list of possible ad identifiers that can be attributed
+  ads: ["sample-campaign-eijb"],
+  // a list of sites where impressions might have been registered
+  source: ["publisher.example"]
+});
+
+
+<xmp class=idl>
+dictionary PrivateAttributionConversionOptions {
+  required unsigned long histogramSize;
+
+  unsigned long lookbackDays = Infinity;
+  PrivateAttributionImpressionType impression;
```

Comment:
```suggestion
  required DOMString aggregationSystemId;
```
Now, I don't think that this is a string that someone will be happy typing often:

```javascript
let whatsit = await privateAttribution.measureConversion({aggregationSystemId: "joe", ...});
```
That's already unwieldy.

Reply:
How about shortening it to `aggregator`?